Adds decorate functionality to module output (#1357)
insightsengineering/teal: vignettes/decorate-module-output.Rmd

Reviewer: please add links to function manuals

Reviewer 2: @pawelru what kind of links are we talking about? Some urls? or you mean roxygen2 links?

Reviewer 2: @pawelru I can add links to our pkgdown vignette, to the reference section

Reviewer: I meant link to the individual function reference page. When you say: "this and this is done by function foo" it would be great to add the link to "foo".

Reviewer 2: yup, added those in here 2dae2b65460d1bfcb347192df1cca3fdcf9a8685

This document outlines the customization options available for modifying the output of `teal` modules. 
You will learn how to use [`teal_transform_module`](https://insightsengineering.github.io/teal/latest-tag/reference/teal_transform_module.html) to modify and enhance the objects created by `teal::modules()`, 
enabling you to tailor the outputs to your specific requirements without rewriting the original module code.

## Decorators

One of ways of adjusting input data or customizing module outputs in `teal` is the usage of `transformators` 
created through `teal_transform_module`. This document will focus on modifying outputs, and for clarity, we will refer 
to these output-modifying objects as decoration objects or decorators.

In below chapter we will present how to create the simplest static decorator with just a server part. Later, we will
present examples on more advanced usage, where decorators contain UI. You will also learn about a convenience 
function that makes it easier to write decorators, called [`make_teal_transform_server`](https://insightsengineering.github.io/teal/latest-tagx/reference/make_teal_transform_server.html). The chapter ends with an 

> example module that utilizes decorators and a snippet that uses this module in `teal` application.

Reviewer: Let's introduce the concept of decoration - this is still not covered and I think it's important to introduce it first. Something like below - feel free to edit.

Let's assume that there is an object `x` of class `"my_class"`. _Decoration_ is a process of changing an object `x` while preserving its class. These changes might add new methods, change data, etc. In simple terms, `foo(x)` is a decoration of `x` if `foo()` preserve input object class. We name `x` as "decorated object" and `foo()` as "decoration function".
In the context of `teal` applications, we always refer to decorating module outputs - plots, tables, etc. In such case, a simple example could be `add_title(x, <plot title>)` function that takes `ggplot2` plot object as `x` argument, apply title change(s) and returns `ggplot2` plot object. It's important to preserve the class so that the "display part" that follows can be run on both decorated and not decorated objects.
The decoration process might be relatively simple (e.g. a single call without additional data) or very complex (e.g. multiple interdependent steps - each requiring additional data coming from dedicated shiny). This functionality is designed to be a power tool for application developers allowing significant output customisation that are not achievable through existing set of module parameters.

Reviewer 2: Thanks @pawelru 
I redacted the text and included as such f329c0eab09a58ad07e9dcf5205377980eeb8019

### Server

> The simplest way to create a decorator is to use `teal_transform_module` without specifying a `ui`.

Reviewer: ```suggestion
The simplest way to create a decorator is to use `teal_transform_module` with just `server` argument provided.
```

This approach adds functionality solely to the server code of the module. 
In the following example, we assume that the module contains an object (of class `ggplot2`) named `plot`. 
We modify the title and x-axis label of plot:

```{r}
library(teal)
static_decorator <- teal_transform_module(
  server = function(id, data) {
    moduleServer(id, function(input, output, session) {
      reactive({
        req(data())
        within(data(), {
          plot <- plot +
            ggtitle("This is title") +
            xlab("x axis")
        })
      })
    })
  }
)
```

To simplify the repetitive elements of writing new decorators 
(e.g., `function(id, data), moduleServer, reactive, within(data, ...)`), 

> you can use the `make_teal_transform_server(expr)` convenience function, which takes a `language` as input:

Reviewer: every time we introduce a new function - let's link it
```suggestion
you can use the [`make_teal_transform_server()`](https://insightsengineering.github.io/teal/latest-tagx/reference/make_teal_transform_server.html) convenience function, which takes a `language` as input:
```

Reviewer 2: Sure, it was linked a couple sentences before, but this does not hurt

```{r}
static_decorator_lang <- teal_transform_module(
  server = make_teal_transform_server(
    expression(
      plot <- plot +
        ggtitle("This is title") +
        xlab("x axis title")
    )
  )
)
```

### UI

To create a decorator with user interactivity, you can incorporate a `shiny` UI module. 
In the example below, the x-axis title is set dynamically via a `textInput`, allowing users to specify their preferred label. 
Note how the input parameters are passed to the [`within`](https://insightsengineering.github.io/teal/latest-tag/reference/teal_data_module.html) function using its `...` argument.

```{r}
interactive_decorator <- teal_transform_module(
  ui = function(id) {
    ns <- NS(id)
    div(
      textInput(ns("x_axis_title"), "X axis title", value = "x axis")
    )
  },
  server = function(id, data) {
    moduleServer(id, function(input, output, session) {
      reactive({
        req(data())
        within(data(),
          {
            plot <- plot +
              ggtitle("This is title") +
              xlab(x_axis_title)
          },
          x_axis_title = input$x_axis_title
        )
      })
    })
  }
)
```

> As in the earlier examples, `make_teal_transform_server(expr)` can simplify the creation of the server component. 

Reviewer: ```suggestion
As in the earlier examples, [`make_teal_transform_server()`](https://insightsengineering.github.io/teal/latest-tagx/reference/make_teal_transform_server.html) can simplify the creation of the server component. 
```

---
title: "Customizing Module Output"
author: "NEST CoreDev"
output:
  rmarkdown::html_vignette:
    toc: true
vignette: >
  %\VignetteIndexEntry{Customizing Module Output}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Introduction

`teal` is a powerful `shiny`-based framework with built-in modules for interactive data analysis. 
This document outlines the customization options available for modifying the output of `teal` modules. 
You will learn how to use [`teal_transform_module`](https://insightsengineering.github.io/teal/latest-tag/reference/teal_transform_module.html) to modify and enhance the objects created by `teal::modules()`, 
enabling you to tailor the outputs to your specific requirements without rewriting the original module code.

## Decorators

> In programming, **decoration** refers to the process of modifying an object while preserving its original class. For instance, given an object `x` of class `"my_class"`, a function `foo(x)` is considered a **decorator function** if it modifies `x` and returns an object that retains the same class. In this context, `x` is referred to as the **decorated object**, and `foo()` is the **decorator function**. Decorators can perform a variety of operations, such as adding new methods or modifying data, while ensuring the object remains compatible with its original usage.

Reviewer: ```suggestion
In programming, **decoration** refers to the process of modifying an object while preserving its original class. For instance, given an object `x` of class `"my_class"`, a function `foo(x)` is considered a **decorator function** if it modifies `x` and returns an object that retains the same class. In this context, `x` is referred to as the **decorated object**, and `foo()` is the **decorator function** or **decorator**. Decorators can perform a variety of operations, such as adding new methods or modifying data, while ensuring the object remains compatible with its original usage.
```

Preserving the object's class during decoration is essential for compatibility. It ensures that the subsequent "display" logic can seamlessly handle both decorated and undecorated objects.

The decoration process can vary in complexity:
- **Simple Decorations**: Single-step modifications, such as a single method call that does not require additional data.
- **Complex Decorations**: Multi-step operations that may involve interdependent transformations, potentially requiring input from dedicated `shiny` UI elements.

This powerful functionality empowers application developers to significantly customize outputs beyond the default capabilities provided by existing module parameters. Decorations allow for advanced modifications, enabling highly tailored and dynamic user experiences in `teal` applications.

## Decorators in `teal`

One of ways of adjusting input data or customizing module outputs in `teal` is the usage of `transformators` 

> created through `teal_transform_module`. This document will focus on modifying outputs, and for clarity, we will refer 
> to these output-modifying objects as decoration objects or decorators.

Reviewer: We already explained these keywords above

```suggestion
created through `teal_transform_module`.
```

In below chapter we will present how to create the simplest static decorator with just a server part. Later, we will
present examples on more advanced usage, where decorators contain UI. You will also learn about a convenience 
function that makes it easier to write decorators, called [`make_teal_transform_server`](https://insightsengineering.github.io/teal/latest-tagx/reference/make_teal_transform_server.html). The chapter ends with an 
example module that utilizes decorators and a snippet that uses this module in `teal` application.

### Server

> The simplest way to create a decorator is to use `teal_transform_module` with just `server` argument provided (i.e. without UI part).

Reviewer: ```suggestion
The simplest way to create a decorator is to use [`teal_transform_module`](https://insightsengineering.github.io/teal/latest-tag/reference/teal_transform_module.html) with just `server` argument provided (i.e. without UI part).
```

```{r}
static_decorator_lang <- teal_transform_module(
  server = make_teal_transform_server(
    expression(
      plot <- plot +
        ggtitle("This is title") +
        xlab("x axis title")
    )
  )
)
```

### UI

To create a decorator with user interactivity, you can add (optional) UI part and use it in server accordingly (i.e. a typical `shiny` module). 
In the example below, the x-axis title is set dynamically via a `textInput`, allowing users to specify their preferred label. 
Note how the input parameters are passed to the [`within`](https://insightsengineering.github.io/teal/latest-tag/reference/teal_data_module.html) function using its `...` argument.

```{r}
interactive_decorator <- teal_transform_module(
  ui = function(id) {
    ns <- NS(id)
    div(
      textInput(ns("x_axis_title"), "X axis title", value = "x axis")
    )
  },
  server = function(id, data) {
    moduleServer(id, function(input, output, session) {
      reactive({
        req(data())
        within(data(),
          {
            plot <- plot +
              ggtitle("This is title") +

>               xlab(x_axis_title)
>           },
>           x_axis_title = input$x_axis_title

Reviewer: Would it work?

```suggestion
              xlab(my_title)
          },
          my_title = input$x_axis_title
```

I just want to differentiate this and following example. In there - we must use names from `input`.

          }), 200
        )

        plot_data_decorated <- srv_teal_transform_data("decorate", data = plot_data, transformators = decorators)

        plot_r <- reactive({
          req(plot_data_decorated())
          plot_data_decorated()[["plot"]]
        })

        output$plot <- renderPlot(plot_r())
        output$text <- renderText({
          teal.code::get_code(req(plot_data_decorated()))
        })
      })
    },
    ui_args = list(decorators = decorators),
    server_args = list(decorators = decorators)
  )
}
```

### Application

```{r}
library(ggplot2)
app <- init(
  data = teal_data(iris = iris, mtcars = mtcars),
  modules = modules(
    tm_decorated_plot("identity"),
    tm_decorated_plot("no-ui", decorators = static_decorator),
    tm_decorated_plot("lang", decorators = static_decorator_lang),
    tm_decorated_plot("interactive", decorators = interactive_decorator),
    tm_decorated_plot("interactive-from lang", decorators = interactive_decorator_lang),
    tm_decorated_plot("from-fun", decorators = gg_xlab_decorator("plot")),
    tm_decorated_plot("failing", decorators = failing_decorator)
  )
)

if (interactive()) {
  shinyApp(app$ui, app$server)
}
```

> ## Multiple Decorators

Reviewer: can we have more general example of multiple (n) decorators?
the current one is for n=2 decorators - not 1, not 0, not 3, etc. - exactly 2

Reviewer 2: I tried to rewrite the code so it works for n decorators. here is the PR, however it is not done yet as this does not produce the final plot https://github.com/insightsengineering/teal/pull/1418

Reviewer 2: Ok, managed to fix the reactivity and merged the vignette to this PR in here https://github.com/insightsengineering/teal/pull/1357/commits/42035b7877c2da21bd2c616ed8d7f7c3dc931482

In the context of `teal` applications, decoration is specifically used to modify module outputs, such as plots or tables. For example, consider a decorator function `add_title(x, <plot_title>)` that takes a `ggplot2` plot object (`x`) as input, applies a title modification, and returns a modified `ggplot2` plot object. This function qualifies as a decorator because it preserves the original class of the input object. Conversely, a function like `create_plot(<data>, <x_axis>, <y_axis>)`, which generates a new plot object, is **not** a decorator, as it produces an output of a different class.

Preserving the object's class during decoration is essential for compatibility. It ensures that the subsequent "display" logic can seamlessly handle both decorated and non-decorated objects.

> The decoration process can vary in complexity:

Reviewer 3: This list doesn't show up nicely on the browser on my computer:

![screenshot of this paragraph](https://github.com/user-attachments/assets/6b643421-5cd6-4a22-a59c-f72b1ca6a9ea)

I think it needs a new line before it:

```suggestion
The decoration process can vary in complexity:

```

Reviewer: Good catch!

My assumption was that this is automatically detected - it seems that it's not

<details>
  <summary>markdownlint results</summary>

  ```
❯ markdownlint vignettes/decorate-module-output.Rmd
vignettes/decorate-module-output.Rmd:15:98 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:15:81 MD013/line-length Line length [Expected: 80; Actual: 98]
vignettes/decorate-module-output.Rmd:16:103 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:16:81 MD013/line-length Line length [Expected: 80; Actual: 103]
vignettes/decorate-module-output.Rmd:17:208 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:17:81 MD013/line-length Line length [Expected: 80; Actual: 208]
vignettes/decorate-module-output.Rmd:18:81 MD013/line-length Line length [Expected: 80; Actual: 108]
vignettes/decorate-module-output.Rmd:22:81 MD013/line-length Line length [Expected: 80; Actual: 587]
vignettes/decorate-module-output.Rmd:24:81 MD013/line-length Line length [Expected: 80; Actual: 597]
vignettes/decorate-module-output.Rmd:26:81 MD013/line-length Line length [Expected: 80; Actual: 190]
vignettes/decorate-module-output.Rmd:29:81 MD013/line-length Line length [Expected: 80; Actual: 120]
vignettes/decorate-module-output.Rmd:29 MD032/blanks-around-lists Lists should be surrounded by blank lines [Context: "- **Simple Decorations**: Sing..."]
vignettes/decorate-module-output.Rmd:30:81 MD013/line-length Line length [Expected: 80; Actual: 161]
vignettes/decorate-module-output.Rmd:32:81 MD013/line-length Line length [Expected: 80; Actual: 290]
vignettes/decorate-module-output.Rmd:38:81 MD013/line-length Line length [Expected: 80; Actual: 222]
vignettes/decorate-module-output.Rmd:39:81 MD013/line-length Line length [Expected: 80; Actual: 272]
vignettes/decorate-module-output.Rmd:41:81 MD013/line-length Line length [Expected: 80; Actual: 136]
vignettes/decorate-module-output.Rmd:45:109 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:45:81 MD013/line-length Line length [Expected: 80; Actual: 109]
vignettes/decorate-module-output.Rmd:48:81 MD013/line-length Line length [Expected: 80; Actual: 116]
vignettes/decorate-module-output.Rmd:49:110 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:49:81 MD013/line-length Line length [Expected: 80; Actual: 110]
vignettes/decorate-module-output.Rmd:50:213 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:50:81 MD013/line-length Line length [Expected: 80; Actual: 213]
vignettes/decorate-module-output.Rmd:51:81 MD013/line-length Line length [Expected: 80; Actual: 98]
vignettes/decorate-module-output.Rmd:55:81 MD013/line-length Line length [Expected: 80; Actual: 227]
vignettes/decorate-module-output.Rmd:56:74 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:57:106 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:57:81 MD013/line-length Line length [Expected: 80; Actual: 106]
vignettes/decorate-module-output.Rmd:78:62 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:79:73 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:80:81 MD013/line-length Line length [Expected: 80; Actual: 202]
vignettes/decorate-module-output.Rmd:96:144 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:96:81 MD013/line-length Line length [Expected: 80; Actual: 144]
vignettes/decorate-module-output.Rmd:97:126 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:97:81 MD013/line-length Line length [Expected: 80; Actual: 126]
vignettes/decorate-module-output.Rmd:98:81 MD013/line-length Line length [Expected: 80; Actual: 180]
vignettes/decorate-module-output.Rmd:126:209 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:126:81 MD013/line-length Line length [Expected: 80; Actual: 209]
vignettes/decorate-module-output.Rmd:127:81 MD013/line-length Line length [Expected: 80; Actual: 193]
vignettes/decorate-module-output.Rmd:149:80 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:150:81 MD013/line-length Line length [Expected: 80; Actual: 123]
vignettes/decorate-module-output.Rmd:151:81 MD013/line-length Line length [Expected: 80; Actual: 133]
vignettes/decorate-module-output.Rmd:152:81 MD013/line-length Line length [Expected: 80; Actual: 119]
vignettes/decorate-module-output.Rmd:182:81 MD013/line-length Line length [Expected: 80; Actual: 123]
vignettes/decorate-module-output.Rmd:183:81 MD013/line-length Line length [Expected: 80; Actual: 134]
vignettes/decorate-module-output.Rmd:205:120 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:205:81 MD013/line-length Line length [Expected: 80; Actual: 120]
vignettes/decorate-module-output.Rmd:206:80 MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
vignettes/decorate-module-output.Rmd:210 MD012/no-multiple-blanks Multiple consecutive blank lines [Expected: 1; Actual: 2]
vignettes/decorate-module-output.Rmd:212:81 MD013/line-length Line length [Expected: 80; Actual: 93]
vignettes/decorate-module-output.Rmd:312 MD024/no-duplicate-heading Multiple headings with the same content [Context: "Example Module"]
vignettes/decorate-module-output.Rmd:314:81 MD013/line-length Line length [Expected: 80; Actual: 308]
vignettes/decorate-module-output.Rmd:406:81 MD013/line-length Line length [Expected: 80; Actual: 87]
vignettes/decorate-module-output.Rmd:421 MD012/no-multiple-blanks Multiple consecutive blank lines [Expected: 1; Actual: 2]
vignettes/decorate-module-output.Rmd:493 MD024/no-duplicate-heading Multiple headings with the same content [Context: "Application"]
  ```
</details>

Your observation is there - among tons of other things :)
[markdownlint](https://github.com/igorshubovych/markdownlint-cli) is designed for `.md` files. I believe that `.Rmd` should have a similar structure (if not identical) - worst case we would have to disable some of the rules. I'm thinking that we should configure our superlinter CI workflow to process `.Rmd` with `markdownlint`. Let me create an issue for this. For the time being - I would recommend to test it manually.

Reviewer: https://github.com/insightsengineering/idr-tasks/issues/712

Reviewer 3: Nice, the heading checks might help to structure better the vignettes or simplify the sections. 
I think you meant to link to https://github.com/insightsengineering/idr-tasks/issues/823

Writing a decorator that applies to any module can be challenging because different modules may use different object names.
It is recommended to create a library of decorator functions that can be adapted to the specific object names used in `teal` modules.
In the following example, focus on the `output_name` parameter to see how decorator can be applied to multiple modules:

```{r}
gg_xlab_decorator <- function(output_name) {
  teal_transform_module(
    ui = function(id) {
      ns <- NS(id)
      div(
        textInput(ns("x_axis_title"), "X axis title", value = "x axis")
      )
    },
    server = function(id, data) {
      moduleServer(id, function(input, output, session) {
        reactive({
          req(data())
          within(data(),
            {
              output_name <- output_name +
                xlab(x_axis_title)
            },
            x_axis_title = input$x_axis_title,
            output_name = as.name(output_name)
          )
        })
      })
    }
  )
}
```

Decorator failures are managed by an internal `teal` mechanism called **trigger on success**, which ensures that the `data`

> object within the module remains intact. If a decorator fails, it will be ignored, and an appropriate error message will be displayed.

Reviewer: isn't it too technical?

Reviewer 2: Like the object will not be changed inside module/teal_data id the decorator produces an error : P

                ggplot2::geom_point()
            },
            dataname = as.name(dataname()),
            x = as.name(x()),
            y = as.name(y())
          )
        })

        plot_data_decorated_no_print <- srv_teal_transform_data(
          "decorate",
          data = plot_data,
          transformators = decorators
        )
        plot_data_decorated <- reactive(
          within(req(plot_data_decorated_no_print()), expr = plot)
        )

        plot_r <- reactive({
          plot_data_decorated()[["plot"]]
        })

        output$plot <- renderPlot(plot_r())
        output$text <- renderText({
          teal.code::get_code(req(plot_data_decorated()))
        })
      })
    },
    ui_args = list(decorators = decorators),
    server_args = list(decorators = decorators)
  )
}
```

### Application

```{r}
library(ggplot2)
app <- init(
  data = teal_data(iris = iris, mtcars = mtcars),
  modules = modules(
    tm_decorated_plot("identity"),
    tm_decorated_plot("no-ui", decorators = list(static_decorator)),
    tm_decorated_plot("lang", decorators = list(static_decorator_lang)),
    tm_decorated_plot("interactive", decorators = list(interactive_decorator)),

>     tm_decorated_plot("interactive-from lang", decorators = list(interactive_decorator_lang)),

Reviewer 3: The interactive-from lang example doesn't change the axis labels (on current branch status): 

https://github.com/user-attachments/assets/703cb08b-f581-40d0-ba6f-06e0b9e89a78

The `interactive_decorator_lang` decorator uses `make_teal_transform_server` which is also used by `static_decorator` ( used in no-ui tab example) but on that example there is no possible change on the axis by the app user. 

Other examples showing how to use decorators to change the axis interactively via function or with an interactive decorator work though.

    tm_decorated_plot("from-fun", decorators = list(gg_xlab_decorator("plot"))),
    tm_decorated_plot("failing", decorators = list(failing_decorator))
  )
)

if (interactive()) {
  shinyApp(app$ui, app$server)
}
```

## Multiple Decorators

### Example Module

It is possible to pass any number of decorators (n) to a module. The example below demonstrates how to handle a dynamic number of decorators, allowing the user to choose which decorator to apply from a list. This makes the module more flexible and capable of accommodating various customization requirements.

> 

Reviewer 3: The code is clear on this regard (see the bottom of the picture):
![image](https://github.com/user-attachments/assets/058e2600-5651-4efc-b461-dab3100fb7b9)
but it might be nice to have a comment explaining that on this case we only expect to keep the last title (as all the three decorators do the same). 

I think it might be better to have one decorator among the others that changes something else, for example the y axis title, this way the composition of the decorators is more clear to the user/developer.

---
title: "Customizing Module Output"
author: "NEST CoreDev"
output:
  rmarkdown::html_vignette:
    toc: true
vignette: >
  %\VignetteIndexEntry{Customizing Module Output}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Introduction

`teal` is a powerful `shiny`-based framework with built-in modules for interactive data analysis. 

> This document outlines the customization options available for modifying the output of `teal` modules. 

Reviewer: one possible way of customization is ... a module parameter(s)
this vignette describes one of possible ways (probably the most powerful and flexible) - a decoration

we need to have a few words what decoration is (and what it is not) and then go to the details

Reviewer 2: 5d09fcf6c4e028cd66b1466dcdd906eec502f135

You will learn how to use `teal_transform_module` to modify and enhance the objects created by `teal` modules, 
enabling you to tailor the outputs to your specific requirements without rewriting the original module code.

Adjusting input data or customizing module outputs in `teal` is accomplished using `transformators` created through 
`teal_transform_module`. This document will focus on modifying outputs, and for clarity, we will refer to these 
output-modifying objects as decoration objects or decorators.

> ## Decorators

Reviewer: Alternatively, you can discuss the concept here in this section.

I am also missing something like: we will start with X then Y and finally Z.

Reviewer 2: 5d09fcf6c4e028cd66b1466dcdd906eec502f135

### Server

The simplest way to create a decorator is to use `teal_transform_module` without specifying a `ui`.

> This approach adds functionality solely to the server code of the `teal` app. 

Reviewer: ```suggestion
This approach adds functionality solely to the server code of the module. 
```

In the following example, we assume that the module contains an object (of class `ggplot2`) named `plot`. 
We modify the title and x-axis label of plot:

```{r}

> library(teal)

Reviewer: you can move library to a separate hidden code chunk upfront

Reviewer 2: I'll keep it as we also display `library(teal)` in different vignettes.

static_decorator <- teal_transform_module(
  server = function(id, data) {
    moduleServer(id, function(input, output, session) {
      reactive({
        req(data())
        within(data(), {
          plot <- plot +
            ggtitle("This is title") +
            xlab("x axis")
        })
      })
    })
  }
)
```

To simplify the repetitive elements of writing new decorators 
(e.g., `function(id, data), moduleServer, reactive, within(data, ...)`), 

> you can use the `make_teal_transform_server(expr)` wrapper, which takes an `expression` as input:

Reviewer: I would probably use a "convenience function" instead of a wrapper. Not a big deal though

Reviewer 2: 0f51ce57af086efc4e52b2bc2fd4e2b46f596399

```{r}

> static_decorator_lang <- teal_transform_module(

Reviewer 2: 6e9ae0c20ac15b95d3d06e98e85b7e7e0c9fef0e

  server = make_teal_transform_server(
    expression(
      plot <- plot +
        ggtitle("This is title") +
        xlab("x axis title")
    )
  )
)
```

### UI

> To create a decorator with user interactivity, you can incorporate a `shiny` UI module. 

Reviewer: ```suggestion
To create a decorator with user interactivity, you can incorporate an UI part and use it in server part (i.e. a typical `shiny` module).
```

Reviewer 2: use UI in server : P ?

Reviewer 2: @pawelru and use it `with server part` instead of `in server part`?

Reviewer: How about this?
```suggestion
To create a decorator with user interactivity, you can add (optional) UI part and use it in server accordingly (i.e. a typical `shiny` module). 
```

      )
    },
    server = function(id, data) {
      moduleServer(id, function(input, output, session) {
        reactive({
          req(data())
          within(data(),
            {
              output_name <- output_name +
                xlab(x_axis_title)
            },
            x_axis_title = input$x_axis_title,
            output_name = as.name(output_name)
          )
        })
      })
    }
  )
}
```

Decorator failures are managed by an internal `teal` mechanism called **trigger on success**, which ensures that the `data`
object within the module remains intact. If a decorator fails, it will be ignored, and an appropriate error message will be displayed.

```{r}
failing_decorator <- teal_transform_module(
  ui = function(id) {
    ns <- NS(id)
    div(
      textInput(ns("x_axis_title"), "X axis title", value = "x axis")
    )
  },
  server = function(id, data) {
    moduleServer(id, function(input, output, session) {
      reactive(stop("This is error"))
    })
  }
)
```

## Decorating Plots

### Example Module

To include decorators in a `teal` module, pass them as arguments (`ui_args` and `server_args`) to the module’s `ui` and 

> `server` components, where they will be used by `ui/srv_teal_transform_module`:

Reviewer: ```suggestion
`server` components, where they will be used by `ui/srv_teal_transform_module`. 

Please find an example module for the sake of this article:
```

```{r}
tm_decorated_plot <- function(label = "module", transformators = list(), decorators = teal_transform_module()) {
  module(
    label = label,
    ui = function(id, decorators) {
      ns <- NS(id)
      div(
        selectInput(ns("dataname"), label = "select dataname", choices = NULL),
        selectInput(ns("x"), label = "select x", choices = NULL),
        selectInput(ns("y"), label = "select y", choices = NULL),
        ui_teal_transform_data(ns("decorate"), transformators = decorators),
        plotOutput(ns("plot")),
        verbatimTextOutput(ns("text"))
      )
    },
    server = function(id, data, decorators) {
      moduleServer(id, function(input, output, session) {
        observeEvent(data(), {
          updateSelectInput(inputId = "dataname", choices = names(data()))
        })

        observeEvent(input$dataname, {
          req(input$dataname)
          updateSelectInput(inputId = "x", choices = colnames(data()[[input$dataname]]))
          updateSelectInput(inputId = "y", choices = colnames(data()[[input$dataname]]))
        })

>         plot_data <- debounce(

Reviewer: why `debounce`?

Reviewer 2: No idea to be honest

Reviewer: Would be great to get rid of this. 

We should aim to have a very simple examples so that the reader can keep focus on what's being described and not on anything extra - such as this one. Also - this gives some false impression that the usage of the `debuonce()` is required for this functionality which is not true.

Author: @m7pr It happens probably because of the reactive which triggers on change of data, updated `input$dataname` and updated `input$x` and `input$y`. Problem here is that any `updateSelectInput` triggers a new reactive cycle.

<img width="539" alt="image" src="https://github.com/user-attachments/assets/bef65955-0aed-45ab-84bb-ab9a482d155d">

Alternative solution could be to `isolate(data())` and `isolate(input$dataname)`, but this would be controversial solution here as ggplot should be recomputed when `data()` changes, when `dataname` changes, and then `x` or `y` changes.

Reviewer 4: I found another aspect of this "extra-reactivity" with the sample app at the bottom. This seems more problematic.

Where modifying the input of another module triggers the decorator's expression of the first module. Illustrated by the circle emojis 🔴 (with decorator) and 🔵 (without decorator)

[Screencast from 2024-11-18 14-12-48.webm](https://github.com/user-attachments/assets/db6445be-71b6-4aad-8136-a1fab8d8cbc6)

The messages inside the mode require the following code added in `example_module`

```r
        # Added to example_module() just before table_data_decorated
        observeEvent(table_data(), logger::log_fatal("table_data reactive {label}"))
        observeEvent(input$dataname, logger::log_fatal("input$dataname reactive {label}"))
        observeEvent(input, logger::log_fatal("input reactive {label}"))
```

<details>

<summary>App code</summary>

```r
pkgload::load_all("teal")

options(teal.log_level = "INFO")

output_decorator <- teal_transform_module(
  label = "a label",
  server = make_teal_transform_server(
    expression({
      logger::log_warn("Expression being called! :: ", paste(ls(), collapse = ", "))
      object <- rev(object)
    })
  )
)

init(
  data = teal.data::teal_data(data1 = iris, data2 = mtcars, data3 = chickwts),
  modules = modules(
    example_module("module 🔴", decorators = output_decorator),
    example_module("module 🔵 (no decorator)", datanames = c("data2", "data3"))
  )
) |> shiny::runApp()
```

</details>

Reviewer 2: @averissimo also found the same in example_module https://github.com/insightsengineering/teal/pull/1357#discussion_r1846519344

Reviewer 4: The problem I reported is fixed after 5b402e268 and was unrelated.

The `debounce` issue can be solved by checking if `x` and `y` exist as colnames of the dataname. 

<details>

<summary>Suggestion here</summary>

```diff
diff --git a/vignettes/decorate-module-output.Rmd b/vignettes/decorate-module-output.Rmd
index 667c8a994..736d4991d 100644
--- a/vignettes/decorate-module-output.Rmd
+++ b/vignettes/decorate-module-output.Rmd
@@ -234,22 +234,29 @@ tm_decorated_plot <- function(label = "module", transformators = list(), decorat
           updateSelectInput(inputId = "x", choices = colnames(data()[[input$dataname]]))
           updateSelectInput(inputId = "y", choices = colnames(data()[[input$dataname]]))
         })
+        
+        dataname <- reactive(req(input$dataname))
+        x <- reactive({
+          req(input$x, input$x %in% colnames(data()[[dataname()]]))
+          input$x
+        })
+        
+        y <- reactive({
+          req(input$y, input$y %in% colnames(data()[[dataname()]]))
+          input$y
+        })
 
-        plot_data <- debounce(
-          reactive({
-            req(input$dataname, input$x, input$y)
-            within(data(),
-              {
-                plot <- ggplot2::ggplot(dataname, ggplot2::aes(x = x, y = y)) +
-                  ggplot2::geom_point()
-              },
-              dataname = as.name(input$dataname),
-              x = as.name(input$x),
-              y = as.name(input$y)
-            )
-          }), 200
-        )
-
+        plot_data <- reactive({
+          req(dataname(), x(), y())
+          
+          within(data(),
+            plot <- ggplot2::ggplot(dataname, ggplot2::aes(x = x, y = y)) + ggplot2::geom_point(),
+            dataname = req(as.name(input$dataname)),
+            x = as.name(x()),
+            y = as.name(y())
+          )
+        })
+        
```

</details>

Reviewer 2: Thanks, @averissimo I think this is still simple enough! For sure less clutter than this obscure debounce. I'll update the vignette's code for all our examples (single decorator case and N decorator case).

Reviewer 2: I updated the vignette with the following suggestion. Pushed the change in here 0003fef718eb10310f3a8be8d95c1db86ec02d8e

Reviewer: Thanks both! This is much better now

        ),
        div(
          id = ns("decorate_table"),
          ui_transform_teal_data(ns("decorate_table"), transformators = decorators$table)
        ),
        plotOutput(ns("plot")),
        tableOutput(ns("table")),
        verbatimTextOutput(ns("text"))
      )
    },
    server = function(id, data, decorators) {
      moduleServer(id, function(input, output, session) {
        observeEvent(data(), {
          updateSelectInput(inputId = "dataname", choices = names(data()))
        })

        observeEvent(input$dataname, {
          req(input$dataname)
          updateSelectInput(inputId = "x", choices = colnames(data()[[input$dataname]]))
          updateSelectInput(inputId = "y", choices = colnames(data()[[input$dataname]]))
        })

        dataname <- reactive(req(input$dataname))
        x <- reactive({
          req(input$x, input$x %in% colnames(data()[[dataname()]]))
          input$x
        })

        y <- reactive({
          req(input$y, input$y %in% colnames(data()[[dataname()]]))
          input$y
        })

        # Generate plot data
        plot_data <- reactive({
          req(dataname(), x(), y())
          within(data(),
            {

>               plot <- ggplot2::ggplot(dataname, aes(x = xvar, y = yvar)) +

Reviewer 4: ```suggestion
              plot <- ggplot2::ggplot(dataname, ggplot2::aes(x = xvar, y = yvar)) +
```

          "decorate_table",
          data = table_data,
          transformators = decorators$table
        )

        output$plot <- renderPlot({
          req(decorated_plot())
          decorated_plot()[["plot"]]
        })

        output$table <- renderTable({
          req(decorated_table())
          decorated_table()[["table_data"]]
        })

        output$text <- renderText({
          plot_code <- teal.code::get_code(req(decorated_plot()))
          table_code <- teal.code::get_code(req(decorated_table()))
          paste("Plot Code:", plot_code, "\nTable Code:", table_code)
        })
      })
    },
    ui_args = list(decorators = decorators),
    server_args = list(decorators = decorators)
  )
}

```

## Example Decorators

1. **Plot Decorator**: Adds a title to the plot.

```{r}
plot_decorator <- teal_transform_module(
  ui = function(id) {
    ns <- NS(id)
    div(
      textInput(ns("plot_title"), "Plot Title", value = "Default Title")
    )
  },
  server = function(id, data) {
    moduleServer(id, function(input, output, session) {
      reactive({
        req(data())
        within(data(),
               {
                 plot <- plot +

>                    ggtitle(ptitle)

Reviewer 4: ```suggestion
                   ggplot2::ggtitle(ptitle)
```

```{r}
table_decorator <- teal_transform_module(
  label = "Decorate plot",
  server = make_teal_transform_server(
    expression({
      table_data$id <- paste0("Row ", seq_len(nrow(table_data)))
      table_data <- table_data[, union("id", colnames(table_data))]

>       colnames(table_data)[1] <- "ID (decorated)"

Reviewer 4: I wanted to make it the first column, but yeah, that simplifies and that's what we need

>       colnames(table_data)[1] <- "ID (decorated)"

Reviewer 4: Changed it in `e2f58e9ce`
